core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSail.java

```diff
@@ -0,0 +1,311 @@
+    Set<Resource> properties = new HashSet<>();
+    List<Statement> subPropertyOfStatements = new ArrayList<>();
+    List<Statement> rangeStatements = new ArrayList<>();
+    List<Statement> domainStatements = new ArrayList<>();
+
+
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        subClassOfStatements = new ArrayList<>();
+        properties = new HashSet<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(FastRdfsForwardChainingSailConnection connection) {
+
+        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new IllegalStateException("Connection already has a lock!");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(FastRdfsForwardChainingSailConnection connection) {
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(FastRdfsForwardChainingSailConnection connection) {
+
+////        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        numberOfThreadsWaitingForWriteLock.incrementAndGet();
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
+
+                if (l != 0) {
+                    long temp = connection.lockStamp;
+                    connection.lockStamp = l;
+                    if (temp != l) {
+//                        System.err.println("readLock: " + temp + " writeLock: " + connection.lockStamp);
+                    }
+
+
+                    return;
+                }
+
+                try {
+                    Thread.sleep(random.nextInt(2));
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
```

Comment:
tryWriteLock will always fail, since th current thread is holding a read lock. It has to be an upgrade attempt.

I initially had a timeout based approach, but that has the downside of killing write transactions even when there are no concurrent write transactions. So in your scenario one, you will end up killing the write transactions for sufficiently long read transactions, even though those scenarios are guaranteed to succeed if the write thread waits until the read transaction is complete.

When it comes to selecting which transactions to cancel, I think I have explained myself a bit poorly. When there are two competing write transactions, one of them doesn't kill the other. What happens is that each transaction will randomly kill itself, just like with a timeout except randomly. So when two transactions enter the upgrade lock code, one of those will randomly succeed in getting the write lock, instead of the code trying to be fair and giving the write lock to the first one that entered (which would require me to keep track of who entered first, and who is still alive).

The termination is not guaranteed, because technically the random number generator could return false indefinitely. I could solve this scenario by having a counter as well, to kill transactions if the random number generator has returned false too many times in a row.